### ISSUE TITLE
Update instance data to no longer use classes as properties

### ISSUE THREAD

**Issue by zednis** [USGCRP/gcis-ontology#12]
There are a number of cases in the instance data where we are incorrectly using classes as properties.  

**This is a pretty big problem and I think fixing it should take precedence over any and all suggested updates to the ontology.**  I think a large number of the suggested changes will become defunct once we fix the instance data.

Several of the suggested changes I have looked at are based more on confusion about how to use the existing model  (and RDF) correctly then there being an identified gap in the current model.  

My hope is that fixing the instance data will resolve most of the current suggested changes.  In cases where we cannot find a suitable property to use we have a well-formed argument for an ontology update.

**Comment by bduggan** [comment]
I agree.  Is there an automated way to discover all of these?
At [ldow2015](http://events.linkeddata.org/ldow2015/), there was a [paper](http://events.linkeddata.org/ldow2015/papers/ldow2015_paper_01.pdf) about fixing domain+range problems in triple stores. maybe there's something we can use from that..

**Comment by zednis** [comment]
I could probably write a script.

**Comment by bduggan** [comment]
On Tuesday, May 26, Stephan Zednik wrote: 

> I could probably write a script.

Sounds great.  This would be great to incorporate
into the unit tests.

Brian

**Comment by justgo129** [comment]
I really like the idea of a script, as it would be too taxing to perform manually.  What would the output of the script be - a list of all the flagged properties/classes arranged according to the the file name for their respective turtle templates?

**Comment by zednis** [comment]
I would probably use python and rdflib to write the script.  It would load the instance data files into a graph and then check each statement to see if the property has type owl:Class or rdfs:Class.

In doing so it would probably lose the line number as it would be looking at the RDF graph in memory.

It could log the statement and the input file.

**Comment by zednis** [comment]
A simpler first script would be to check for any statement properties that start with a capital letter.  That script would not be as thorough but would be faster to write and could spit out the line number.

note - RDF conventions are that classes start with capital letters and properties start with lowercase letters.  This script will catch cases where the properties and classes were named according to conventions.

**Comment by bduggan** [comment]
I was thinking maybe there'd be a quick query against the [endpoint](http://data.globalchange.gov/sparql)...would save you the trouble of scraping everything...

**Comment by zednis** [comment]
@bduggan  Good idea!  Are the ontologies also loaded into the endpoint?

**Comment by bduggan** [comment]
No, do I need to do that from the backend?  By the way, [here](https://github.com/USGCRP/gcis-rdf/blob/master/import-gcis) is the script that drops and recreates the triple store.

**Comment by zednis** [comment]
If we want to check the type of the property used in a statement the definition of the URI used as the property will have to be loaded in the triplestore.

The following query will print out a statement where a class is used as a property but only if the type of ?p is saved in the triplestore.

```
SELECT ?s ?p ?o
WHERE {
  ?s ?p ?o .
  { ?p a owl:Class } UNION {?p a rdfs:Class }  .
}
```

This means we will need to load the RDF for dublin core terms, gcis, prov, and any sweet ontologies we are using into the triplestore.

**Comment by bduggan** [comment]
Okay.  the  list is [here](https://github.com/USGCRP/gcis/blob/master/lib/Tuba/Plugin/TubaHelpers.pm#L501).  Loading them is straightforward, but finding the turtle for the ontologies will take some investigation. (edit: updated link)

**Comment by zednis** [comment]
Let's do the simple script that scraps content first.  That should give us a good first pass and it doesn't require us to find and load all the schema graphs.  It should also be easy to write in perl.

by convention:
- ex:foo ==> property
- ex:Foo ==> class

**Comment by bduggan** [comment]
[Here](https://gist.github.com/bduggan/f34bc07d1afdf3c63055)'s a start.

**Comment by justgo129** [comment]
@bduggan @zednis could you please rerun when you get a chance so that I can have a better understanding of the remaining extent of this issue?  Thanks.

**Comment by bduggan** [comment]
On Thursday, July 16, justgo129 wrote: 

> @bduggan @zednis could you please rerun when you get a chance so that I can have a better understanding of the remaining extent of this issue?  Thanks.

@justgo129, you can run it: let me know if you need help
running a find command from a bash prompt.

Brian

**Comment by justgo129** [comment]
Sure, I'll run it after we complete this round of pull requests.  Is there a way to modify the script so that classes don't show up if they are used correctly within the turtle?  Alternatively stated, as the code runs now, line 17 of:

https://github.com/USGCRP/gcis/blob/1db0066ddb5ae3a3f6955ed7d31da65b00d5ddca/lib/Tuba/files/templates/article/object.ttl.tut

would show up on the list of classes in gcis.  However, the class is used correctly in this case.

**Comment by bduggan** [comment]
Maybe, but it doesn't make much sense to me that you'd want to do this since you can just eyeball the results.

**Comment by justgo129** [comment]
I just did a search to identify areas in which this is still an issue:
https://github.com/USGCRP/gcis/blob/master/lib/Tuba/files/templates/model_run/object.ttl.tut
  (model_run->Scenario)

https://github.com/USGCRP/gcis/blob/master/lib/Tuba/files/templates/scenario/object.ttl.tut
   (scenario->Model_run)

https://github.com/USGCRP/gcis/blob/master/lib/Tuba/files/templates/project/object.ttl.tut
   (project->Model)

While we are resolving many of these through other pull requests (e.g. https://github.com/USGCRP/gcis-ontology/issues/17) I'm still identifying these to facilitate the process of closing out this ticket.

**Comment by zednis** [comment]
Thanks @justgo129 

(ModelRunOutput->Scenario)

ModelRunOutput and Scenario are both entities.  Entity->Entity relationships in PROV are derivation relationships.  If a scenario is conceptually an input into the (not necessarily represented in RDF) model run activity, then a derivation relationship between the ModelRunOutput and Scenario is semantically accurate.  If not, then we should not use a PROV sub-property here.

```
gcis:derivedFromScenario a owl:ObjectProperty
    rdfs:label "derived from scenario" ;
    rdfs:domain gcis:ModelRunOutput ;
    rdfs:range gcis:Scenario ;
    rdfs:subPropertyOf prov:wasDerivedFrom .
```

Alternatively we could use `gcis:describesScenario` but based on the comment I am not sure this would align with the original intent of the property.

(Scenario->ModelRunOutput)

If a derivation relationship holds between ModelRunOutput and Scenario then this relationship should be `prov:hadDerivation` or some sub-property of it.

```
gcis:derivedModelRunOutputs a owl:ObjectProperty
    rdfs:label "derived model run outputs" ;
    rdfs:domain gcis:Scenario ;
    rdfs:range gcis:ModelRunOutput ;
    owl:inverseOf gcis:derivedFromScenario ;
    rdfs:subPropertyOf prov:hadDerivation .
```

Alternatively we could use `gcis:scenarioDescribedIn` but based on the comment I am not sure this would align with the original intent of the property.

(Project->Model)

We could use `prov:contributed` directly (inverse of `prov:wasAttributedTo`) or we could define our own property that is a subproperty of `prov:contributed` and specific to models.  

```
gcis:contributedModel a owl:ObjectProperty ;
    rdfs:label "contributed model" ;
    rdfs:domain gcis:Project ;
    rdfs:range gcis:Model ;
    rdfs:subPropertyOf prov:contributed .
```

**Comment by rewolfe** [comment]
It sounds like we are trying to say that a ModelRunOuput is basedOn a
Scenario.

On Tue, Sep 15, 2015 at 5:45 PM, Stephan Zednik notifications@github.com
wrote:

> Thanks @justgo129 https://github.com/justgo129
> 
> (ModelRunOutput->Scenario)
> 
> ModelRunOutput and Scenario are both entities. Entity->Entity
> relationships in PROV are derivation relationships. If a scenario is
> conceptually an input into the (not necessarily represented in RDF) model
> run activity, then a derivation relationship between the ModelRunOutput and
> Scenario is semantically accurate. If not, then we should not use a PROV
> sub-property here.
> 
> gcis:derivedFromScenario a owl:ObjectProperty
>     rdfs:label "derived from scenario" ;
>     rdfs:domain gcis:ModelRunOutput ;
>     rdfs:range gcis:Scenario ;
>     rdfs:subPropertyOf prov:wasDerivedFrom .
> 
> Alternatively we could use gcis:describesScenario but based on the
> comment I am not sure this would align with the original intent of the
> property.
> 
> (Scenario->ModelRunOutput)
> 
> If a derivation relationship holds between ModelRunOutput and Scenario
> then this relationship should be prov:hadDerivation or some sub-property
> of it.
> 
> gcis:derivedModelRunOutputs a owl:ObjectProperty
>     rdfs:label "derived model run outputs" ;
>     rdfs:domain gcis:Scenario ;
>     rdfs:range gcis:ModelRunOutput ;
>     owl:inverseOf gcis:derivedFromScenario ;
>     rdfs:subPropertyOf prov:hadDerivation .
> 
> Alternatively we could use gcis:scenarioDescribedIn but based on the
> comment I am not sure this would align with the original intent of the
> property.
> 
> (Project->Model)
> 
> We could use prov:contributed directly (inverse of prov:wasAttributedTo)
> or we could define our own property that is a subproperty of
> prov:contributed and specific to models.
> 
> gcis:contributedModel a owl:ObjectProperty ;
>     rdfs:label "contributed model" ;
>     rdfs:domain gcis:Project ;
>     rdfs:range gcis:Model ;
>     rdfs:subPropertyOf prov:contributed .
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/USGCRP/gcis-ontology/issues/12#issuecomment-140555900
> .

## 

Robert Wolfe, NASA GSFC @ USGCRP, o: 202-419-3470, m: 301-257-6966

**Comment by zednis** [comment]
That sounds reasonable.

(ModelRunOutput -> Scenario) `gcis:basedOn`

(Scenario -> ModelRunOutput) `gcis:basisFor`

Do you think these properties should be sub-properties of the PROV wasDerivedFrom and hadDerivation properties?

**Comment by justgo129** [comment]
I think so, @zednis.

**Comment by rewolfe** [comment]
@zednis - Sounds good.

**Comment by zednis** [comment]
Great.  I am on PTO.  Marshall can create a pull request with the proposed changes.

**Comment by justgo129** [comment]
Great.

**Comment by justgo129** [comment]
@xgmachina please feel free to proceed.

**Comment by justgo129** [comment]
Great.  Now that #159 has been merged, I think we'll just need to add gcis:basedOn and gcis:basisFor per the comment of @zednis that begins with "That sounds reasonable."  We should subsequently be able to close #12.

**Comment by justgo129** [comment]
@zednis or @xgmachina when you get a chance, could you please add gcis:basedOn and gcis:basisFor to the ontology, after which we can close this ticket?  Thanks a million.

**Comment by zednis** [comment]
@justgo129 will do

**Comment by zednis** [comment]
It looks like @xgmachina created the properties, but gave them different names.  Should I rename them?

https://github.com/USGCRP/gcis-ontology/pull/159/files

```
gcis:derivedFromScenario a owl:ObjectProperty ;
        rdfs:label "derived from scenario" ;
    rdfs:comment "A model run output was derived from one or more scenarios that was used in the model run." ;
        rdfs:domain gcis:ModelRunOutput ;
        rdfs:range gcis:Scenario ;
        rdfs:subPropertyOf prov:wasDerivedFrom .

gcis:derivedModelRunOutputs a owl:ObjectProperty ;
        rdfs:label "derived model run outputs" ;
    rdfs:comment "A scenario was used in a model run and derived one or more model run outputs." ;
        rdfs:domain gcis:Scenario ;
        rdfs:range gcis:ModelRunOutput ;
        owl:inverseOf gcis:derivedFromScenario ;
        rdfs:subPropertyOf prov:hadDerivation .
```

**Comment by justgo129** [comment]
I like both approaches, but have a slight preference for yours, Stephan.  @zednis feel free to rename.

**Comment by zednis** [comment]
@justgo129 after thinking about it I think I have a slight preference for the names @xgmachina used.  `gcis:basedOn` and `gcis:basisFor` are pretty generic for the specific domain and range of the property and could lead to the property being misused by someone who does not take a careful look at the property domain and range.

I would suggest leaving as is.  I also think this ticket - which was originally about resolving cases where we were using non-properties as properties - is ready to be closed.  If we find any further cases of non-properties being used as properties in the instance data we can create a specific ticket for that occurrence.

**Comment by justgo129** [comment]
Sounds good.  Closed #12 due partially to merged #159.
